src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,852 @@
+
+        n_initial_params = sum(
+            list(map(lambda b: b.get_free_parameters(), initial_blocks))
+        )
+
+        n_layer_params = (
+            sum(list(map(lambda b: b.get_free_parameters(), layer_blocks)))
+            * self.n_layers
+        )
+
+        return n_initial_params + n_layer_params
+
+    def construct_circuit(self, angles):
+        """
+        Construct a circuit by specifying values for each
+        free parameter.
+
+        Parameters
+        ----------
+        angles: list of float
+            A list of dimension (n,) for n free parameters in the circuit
+
+        Returns
+        -------
+        circ: QuantumCircuit
+        """
+        circ = QubitCircuit(self.n_qubits)
+        circ.user_gates = self.user_gates
+        i = 0
+        for layer_num in range(self.n_layers):
+            for block in self.blocks:
+                if block.initial and layer_num > 0:
+                    continue
+                if block.is_native_gate:
+                    circ.add_gate(block.operator, targets=block.targets)
+                else:
+                    n = block.get_free_parameters()
+                    circ.add_gate(
+                        block.name,
+                        targets=list(range(self.n_qubits)),
+                        arg_value=angles[i : i + n] if n > 0 else None,
+                    )
+                    i += n
+        return circ
+
+    def get_initial_state(self):
+        """
+        Returns
+        -------
+        initial_state: Qobj
+            Initial circuit state
+        """
+        initial_state = basis(2, 0)
+        for _ in range(self.n_qubits - 1):
+            initial_state = tensor(initial_state, basis(2, 0))
+        return initial_state
+
+    def get_final_state(self, angles):
+        """
+        Evaluate the circuit by specifying each circuit parameter
+
+        Parameters
+        ----------
+        angles: list of float
+            A list of dimension (n,) for n free parameters in the circuit
+
+        Returns
+        -------
+        final_state: Qobj
+            Final state of the circuit after evaluation
+        """
+        circ = self.construct_circuit(angles)
+        initial_state = self.get_initial_state()
+        final_state = circ.run(initial_state)
+        return final_state
+
+    def _sample_bitstring_from_state(self, state):
+        """
+        Use probability amplitudes from state after measurement
+        in computational basis to sample a bitstring.
+
+        Parameters
+        ----------
+        state: Qobj
+
+        Returns
+        -------
+        bitstring: str
+            Formatted binary string with length corresponding to
+            dimension of input state
+
+        E.g. the state 1/sqrt(2) * (|0> + |1>)
+        would return 0 and 1 with equal probability.
+        """
+        n_qubits = int(np.log2(state.shape[0]))
+        outcome_indices = list(range(2 ** n_qubits))
```

Comment:
In the new black styling released a few days ago, it is recommended to omit the space around `**`. That's why the lint test is failing. You will get it right with an updated version of black.